Patch for Koha to use RFIDhub

diff --git a/koha-tmpl/intranet-tmpl/prog/en/modules/admin/preferences/circulation.pref b/koha-tmpl/intranet-tmpl/prog/en/modules/admin/preferences/circulation.pref
--- a/koha-tmpl/intranet-tmpl/prog/en/modules/admin/preferences/circulation.pref
+++ b/koha-tmpl/intranet-tmpl/prog/en/modules/admin/preferences/circulation.pref
@@ -751,6 +751,15 @@ Circulation:
                   yes: Show
                   no: "Don't show"
             - "the print receipt popup dialog when self checkout is finished"
+    RFID:
+        -
+            - pref: UseRFIDHub
+              choices:
+                  yes: Use
+                  no: "Don't use"
+            - the RFIDHub located at
+            - pref: RFIDHubURL
+            - "Please note: You need to set up an RFID hub for this to work!"
     Course Reserves:
         -
             - pref: UseCourseReserves
diff --git a/koha-tmpl/intranet-tmpl/prog/en/modules/circ/circulation.tt b/koha-tmpl/intranet-tmpl/prog/en/modules/circ/circulation.tt
--- a/koha-tmpl/intranet-tmpl/prog/en/modules/circ/circulation.tt
+++ b/koha-tmpl/intranet-tmpl/prog/en/modules/circ/circulation.tt
@@ -121,6 +121,7 @@ $(document).ready(function() {

 //]]>
 </script>
+[% IF Koha.Preference( 'UseRFIDHub' ) == '1' %][% INCLUDE 'rfidhub.inc' %][% END %]
 </head>
 <body id="circ_circulation" class="circ">

@@ -868,7 +869,7 @@ No patron matched <span class="ex">[% message %]</span>
 </div>
 </div>

-<div class="yui-g"><div id="patronlists" class="toptabs">
+<div class="yui-g">[% IF Koha.Preference( 'UseRFIDHub' ) == '1' %]<div id="rfidlist"></div>[% END %]<div id="patronlists" class="toptabs">

 <ul>
     <li>
diff --git a/koha-tmpl/intranet-tmpl/prog/en/modules/circ/circulation-home.tt b/koha-tmpl/intranet-tmpl/prog/en/modules/circ/circulation-home.tt
--- a/koha-tmpl/intranet-tmpl/prog/en/modules/circ/circulation-home.tt
+++ b/koha-tmpl/intranet-tmpl/prog/en/modules/circ/circulation-home.tt
@@ -2,6 +2,7 @@
 [% INCLUDE 'doc-head-open.inc' %]
 <title>Koha &rsaquo; Circulation</title>
 [% INCLUDE 'doc-head-close.inc' %]
+[% IF Koha.Preference( 'UseRFIDHub' ) == '1' %][% INCLUDE 'rfidhub.inc' %][% END %]
 </head>
 <body id="circ_circulation-home" class="circ">
 [% INCLUDE 'header.inc' %]
diff --git a/koha-tmpl/intranet-tmpl/prog/en/modules/circ/returns.tt b/koha-tmpl/intranet-tmpl/prog/en/modules/circ/returns.tt
--- a/koha-tmpl/intranet-tmpl/prog/en/modules/circ/returns.tt
+++ b/koha-tmpl/intranet-tmpl/prog/en/modules/circ/returns.tt
@@ -146,6 +146,7 @@ $(document).ready(function () {
 });
 //]]>
 </script>
+[% IF Koha.Preference( 'UseRFIDHub' ) == '1' %][% INCLUDE 'rfidhub.inc' %][% END %]
 </head>
 <body id="circ_returns" class="circ">
 <span class="audio-alert-success"></span>
@@ -587,6 +588,7 @@ $(document).ready(function () {
     </form>
 </div>

+[% IF Koha.Preference( 'UseRFIDHub' ) == '1' %]<div id="rfidlist"></div>[% END %]
 [% IF ( riloop ) %]
     <h2>Checked-in items</h2>
     <table id="checkedintable">
diff --git a/koha-tmpl/intranet-tmpl/prog/en/includes/rfidhub.inc b/koha-tmpl/intranet-tmpl/prog/en/includes/rfidhub.inc
new file mode 100644
index 0000000..ff48de6
--- /dev/null
+++ b/koha-tmpl/intranet-tmpl/prog/en/includes/rfidhub.inc
@@ -0,0 +1,318 @@
+[% USE Koha %]
+<style type="text/css">
+/*
+ * CSS to draw overlay, dialog box and RFID table output
+ */
+.overlay    { position: fixed; top:0; left:0; width:100%; height: 100%; z-index:99;
+              background-color:#000; background: rgba(0,0,0,0.8); /*pointer-events:none;*/ }
+
+#RFIDStatus { clear: left; margin-left: 10px; }
+#RFIDStatus .btn { margin-right: 10px; background-image: none; background-color: orange;}
+
+#writeRFIDDialog {
+  width: 430px; margin: 60px auto; padding: 60px 30px;
+  background: #c9d0de; border: 1px solid #e1e1e1;
+  border-radius: 5px; -moz-border-radius: 5px; -webkit-border-radius: 5px;
+  -moz-box-shadow: 0px 0px 8px #444;
+  -webkit-box-shadow: 0px 0px 8px #444;
+}
+
+#writeRFIDDialog label {
+  float: left; clear: left; margin: 11px 20px 0 0; width: 95px;
+  text-align: right; font-size: 16px; color: #445668;
+  text-transform: uppercase; text-shadow: 0px 1px 0px #f2f2f2;
+}
+
+#writeRFIDDialog input {
+  width: 260px; height: 35px; padding: 5px 20px 0px 20px; margin: 0 0 20px 0;
+  background: #5E768D;
+  border-radius: 5px; -moz-border-radius: 5px; -webkit-border-radius: 5px;
+  -moz-box-shadow: 0px 1px 0px #f2f2f2;-webkit-box-shadow: 0px 1px 0px #f2f2f2;
+  font-family: sans-serif; font-size: 16px; color: #f2f2f2; text-shadow: 0px -1px 0px #334f71;
+}
+
+#writeRFIDDialog a { float: right;}
+#writeRFIDDialog input::-webkit-input-placeholder  {
+      color: #a1b2c3; text-shadow: 0px -1px 0px #38506b;
+}
+#writeRFIDDialog input:-moz-placeholder {
+      color: #a1b2c3; text-shadow: 0px -1px 0px #38506b;
+}
+</style>
+<style type="text/css" media="print">
+  body {
+    visibility: hidden;
+  }
+  table {
+    visibility: visible;
+    position: absolute;
+    left: 0;
+    top: 0;
+  }
+  /* reset table for slip printing */
+  table, tbody, tfoot, thead, tr, th, td {
+    margin: 0;
+    border: 0 !important;
+    font-size: 8px !important;
+    width: auto;
+  }
+  /* display only columns 1-6 and 10 */
+  th, td {
+    display: none;
+  }
+  th:nth-of-type(n+1):nth-of-type(-n+2),th:nth-of-type(n+6) {
+      display: table-cell;
+  }
+  td:nth-of-type(n+1):nth-of-type(-n+2),td:nth-of-type(n+6) {
+      display: table-cell;
+  }
+</style>
+<script type="text/javascript">
+var rfidhuburl = "[% Koha.Preference( 'RFIDHubURL' ) %]";
+/*
+ * JavaScript to inject Table for RFID items
+ * connects to rfid hub with Websocket
+ * creates <div> with <table>
+ * inserts table to #patronlists and adds highlighted items as they arrive from websocket
+ */
+$(document).ready(function() {
+  ws=new WebSocket(rfidhuburl);
+  branch = document.getElementById("logged-in-branch-code") ? document.getElementById("logged-in-branch-code").innerHTML : "ukjent";
+  state = $("body").attr("id");
+  // activate RFID Modal
+  $("#menu ul").append('<li><a href="#" onclick="activateWrite();">Preg RFID</a></li>');
+  $("div.gradient").append('<div id="RFIDStatus"><a href="#" class="btn btn-small" onclick="reconnectHUB();">RFID</a><span id="RFIDStatusMessage"></span></div>');
+  ws.onopen = function() {
+    console.log("connected");
+
+    ws.onmessage = function(resp) {
+      var data = JSON.parse(resp.data);
+      // Major Error handling
+      if (data.RFIDError) { // RFIDhun cannot access RFID
+        $('#RFIDStatus .btn').css("backgroundColor", "red");
+        $('#RFIDStatusMessage').text("ERROR connecting to RFID, please press button or refresh Browser to try again!");
+      } else if (data.SIPError) { // RFIDhub cannot access SIP Server
+        $('#RFIDStatus .btn').css("backgroundColor", "red");
+        $('#RFIDStatusMessage').text("ERROR connecting to SIP Server, please ask Zarathustra for help or refresh browser!");
+      } else if (data.UserError) { // User initiated error
+        $('#RFIDStatus .btn').css("backgroundColor", "red");
+        $('#RFIDStatusMessage').text(data.ErrorMessage);
+      } else {
+        // Handle messages
+        switch (data.Action) {
+          case "CONNECT":
+            console.log("Connected to RFIDhub");
+            $('#RFIDStatus .btn').css("backgroundColor", "lightgreen");
+            if (state == "circ_returns") {
+              ws.send(JSON.stringify({Action: "CHECKIN", Branch: branch}));
+            } else if (state == "circ_circulation") {
+              ws.send(JSON.stringify({Action: "CHECKOUT", Branch: branch, Patron: "[% borrower.cardnumber %]"}));
+            }
+            break;
+          case "DISCONNECT":
+            console.log("Disconnected from RFIDhub");
+            $('#RFIDStatus .btn').css("backgroundColor", "red");
+            $('#RFIDStatusMessage').text("Disconnected from RFIDHub, please refresh Browser to try to reconnect!");
+            break;
+          case "CHECKIN":
+            console.log(" <- "+resp.data);
+            handleItem(data);
+            break;
+          case "CHECKOUT":
+            console.log(" <- "+resp.data);
+            handleItem(data);
+            break;
+          case "ITEM-INFO":
+            console.log(" <- "+resp.data);
+            if (data.Item.Unknown) { // Failed to Look up Item
+              $('#itemStatusDiv').toggleClass('hidden', false);
+              $('#itemStatus').text("Could not find Item with barcode: "+data.Item.Barcode);
+              $('#itemStatus').attr("style", "background-color: pink");
+            } else {
+              displayItemInfo(data.Item)
+            }
+            break;
+          case "WRITE":
+            console.log(" <- "+resp.data);
+            $('#itemStatusDiv').toggleClass('hidden', false);
+            if (data.Item.WriteFailed || data.Item.TagCountFailed) { // Failed to Write
+              $('#itemStatus').text("ERROR: "+data.Item.Status);
+              $('#itemStatus').attr("style", "background-color: pink");
+            } else {
+              $('#itemStatus').text("SUCCESS writing tag!");
+              $('#itemStatus').attr("style", "background-color: lightgreen");
+            }
+            break;
+          }
+      }
+    }
+  }
+  ws.onclose = function() {
+    console.log("disconnected");
+    $('#RFIDStatus .btn').css("backgroundColor", "orange");
+  }
+  ws.onerror = function() {
+    console.log("No connection to RFIDhub!");
+    $('#RFIDStatus .btn').css("backgroundColor", "red");
+    $('#RFIDStatusMessage').text("ERROR connecting to RFIDhub, please refresh Browser to try again!");
+  }
+});
+
+function reconnectHUB() {
+  window.location.reload();
+}
+
+function handleItem(msg) {
+
+  var rfiddiv = getRfidDiv();
+
+  // create table, headers and tbody if not exists
+  if ($('#rfiditems').length == 0) {
+    var table = $('<table/>', { id: 'rfiditems' });
+    table.append('<thead>' +
+      '<tr>' +
+        '<th scope="col">Dato</th>' +
+        '<th scope="col">Tittel</th>' +
+        '<th scope="col">Forfatter</th>' +
+        '<th scope="col">Lånt ut fra</th>' +
+        '<th scope="col">Hyllesignatur</th>' +
+        '<th scope="col">Strekkode</th>' +
+        '<th scope="col">Beskjed</th>');
+    table.append('<tbody/>');
+
+    rfiddiv.append(table);
+  }
+
+  // create row
+  var row = $('<tr>').attr('id', msg.Item.Barcode)
+    .append('<td><span title="dato">'+msg.Item.Date+'</span></td>')
+    .append('<td><strong>'+msg.Item.Label+'</strong></a></td>')
+    .append($('<td>'))
+    .append($('<td>'))
+    .append($('<td>'))
+    .append('<td><a href="/cgi-bin/koha/catalogue/search.pl?q='+
+        msg.Item.Barcode+'">'+msg.Item.Barcode+'</a></td>')
+    .append('<td>'+msg.Item.Status);
+
+  // if row with id already exists: update - else prepend
+  if($("#" + msg.Item.Barcode).length == 0) {
+    rfiddiv.find('tbody').prepend(row);
+  } else {
+    rfiddiv.find($("#" + msg.Item.Barcode).closest('tr')).replaceWith(row);
+  }
+
+  // highlight according to success or failure
+  if(msg.Item.TransactionFailed) {
+    row.children('td').css("backgroundColor", "pink");
+  } else if(msg.Item.AlarmOnFailed || msg.Item.AlarmOffFailed) {
+    row.children('td').css("backgroundColor", "pink");
+    row.append('<td><a href="#" class="btn btn-small" onclick="return retryAlarm();">Retry</a></td>');
+  } else {
+    row.children('td').effect("highlight", {color: "lightgreen"}, 3000);
+  }
+  // popup print dialog for transfers
+  if(msg.Item.Transfer) {
+    window.open('transfer-slip.pl?barcode='+msg.Item.Barcode+'&amp;branchcode='+msg.Item.Transfer+'&amp;op=slip', 'popup', 'width=600,height=400,resizable=1,toolbar=0,scrollbars=1,top');
+  }
+
+  // popup print dialog for hold to wait on current branch
+  if (msg.Item.Hold) {
+    window.open('hold-transfer-slip.pl?borrowernumber=' + msg.Item.Borrowernr + '&amp;biblionumber=' + msg.Item.Biblionr,  'popup', 'width=600,height=400,resizable=1,toolbar=0,scrollbars=1,top');
+  }
+}
+
+function retryAlarm() {
+  if (state == "circ_returns") {
+    ws.send(JSON.stringify({Action: "RETRY-ALARM-ON"}));
+  } else if (state == "circ_circulation") {
+    ws.send(JSON.stringify({Action: "RETRY-ALARM-OFF"}));
+  }
+}
+
+/*
+ * WRITE RFID
+ */
+function activateWrite() {
+  // End Scan session first and activate Write Modal Dialog
+  ws.send(JSON.stringify({Action: "END" }));
+  buildWriteRFIDModal();
+  $("#writeRFIDModal").toggleClass('hidden', false);
+}
+
+function deactivateWrite() {
+  // End Scan session first and activate Write Modal Dialog
+  ws.send(JSON.stringify({Action: "CHECKOUT", Patron: "[% borrower.cardnumber %]" }));
+  $("#writeRFIDModal").toggleClass('hidden', true);
+  return true;
+}
+
+function buildWriteRFIDModal() {
+  var writeRFIDModal = getWriteRFIDModal();
+  writeRFIDModal.html('<div id="writeRFIDDialog">'+
+  '<h3>Write RFID Tag</h3>'+
+  '<label for="name">barcode:</label>'+
+  '<input type="text" id="barcodeInput" placeholder="barcode here">'+
+  '<div id="itemInfoDiv" class="hidden">'+
+  '<label for="name">Label:</label>'+
+  '<input type="text" id="itemLabel" disabled="disabled">'+
+  '<label for="name">NumTags:</label>'+
+  '<input type="number" id="itemTags">'+
+  '</div>'+
+  '<div id="itemStatusDiv" class="hidden">'+
+  '<p id="itemStatus">&nbsp;</p>'+
+  '</div>'+
+  '<div id="writeRFIDButtons">'+
+  '<a href="#" id="writeButton" class="hidden btn btn-small" onclick="return writeItem();">Preg brikker</a></div>'+
+  '<a href="#" class="btn btn-small" onclick="return lookupItem();">Søk opp bok</a>'+
+  '<a href="#" class="btn btn-small" onclick="return deactivateWrite();">Avslutt</a></div>');
+}
+
+function lookupItem() {
+  // Request item info
+  ws.send(JSON.stringify({Action: "ITEM-INFO", Item: { Barcode: $('#barcodeInput').val() } }));
+  return true;
+}
+
+function writeItem() {
+  var msg = JSON.stringify({Action: "WRITE", Item: { Barcode: $('#barcodeInput').val(), NumTags: parseInt($('#itemTags').val()) } });
+  console.log(msg);
+  ws.send(msg);
+  return true;
+}
+
+function displayItemInfo(item) {
+  console.log(item);
+  $("#itemLabel").attr("value", item.Label);
+  $("#itemTags").attr("value", item.NumTags);
+  $("#itemInfoDiv").toggleClass('hidden', false);
+  $("#writeButton").toggleClass('hidden', false);
+}
+
+function getRfidDiv() {
+  // Create div with id rfidlist if not exists
+  if ($('#rfidlist').length == 0) {
+    var rfiddiv = $('<div/>', {
+      id: 'rfidlist',
+      class: 'toptabs ui-tabs ui-widget ui-widget-content ui-corner-all'
+    }).prependTo('#patronlists');
+  } else {
+    var rfiddiv = $('#rfidlist');
+  }
+
+  return rfiddiv;
+}
+
+function getWriteRFIDModal() {
+  // Create modal div with id writeRFIDModal if not exists
+  if ($('#writeRFIDModal').length == 0) {
+    var writeRFIDModal = $('<div/>', {
+      id: 'writeRFIDModal',
+      class: 'overlay'
+    }).prependTo('#bd');
+  } else {
+    var writeRFIDModal = $('#writeRFIDModal');
+  }
+  return writeRFIDModal;
+}
+
+</script>
diff --git a/installer/data/mysql/atomicupdate/bug_11858-add_rfidhub_syspref.sql b/installer/data/mysql/atomicupdate/bug_11858-add_rfidhub_syspref.sql
new file mode 100644
index 0000000..98aee33
--- /dev/null
+++ b/installer/data/mysql/atomicupdate/bug_11858-add_rfidhub_syspref.sql
@@ -0,0 +1 @@
+INSERT IGNORE INTO systempreferences (variable,value,options,explanation,type) VALUES
+('UseRFIDHub','1','','Enable RFID hub','YesNo'),
+('RFIDHubURL','','','URL for the RFID hub.,'Free');
+
