diff --git a/Koha/REST/V1.pm b/Koha/REST/V1.pm
--- a/Koha/REST/V1.pm
+++ b/Koha/REST/V1.pm
@@ -76,6 +76,11 @@ sub authenticate_api_request {
     my ($status, $sessionID) = check_cookie_auth(
                                             $cookie, undef,
                                             { remote_addr => $remote_addr });
+
+    # Deichman mod: allow request to verify patron PIN from overview-page
+    # TODO restrict to domain koha.deichman.no or whatever when subdomain-names are fixed
+    $c->res->headers->header('Access-Control-Allow-Origin' => '*');
+
     if ($status eq "ok") {
         $session = get_session($sessionID);
         $user = Koha::Patrons->find($session->param('number'));
