diff --git a/members/memberentry.pl b/members/memberentry.pl
--- a/members/memberentry.pl
+++ b/members/memberentry.pl
@@ -305,7 +305,7 @@ if ($op eq 'save' || $op eq 'insert'){
     if ( $newdata{dateofbirth} ) {
         my $age = GetAge($newdata{dateofbirth});
         my $borrowercategory = Koha::Patron::Categories->find($newdata{categorycode});
-        my ($low,$high) = ($borrowercategory->dateofbirthrequired, $borrowercategory->upperagelimit);
+        my ($low,$high) = ($borrowercategory->dateofbirthrequired, $borrowercategory->upperagelimit) if $borrowercategory;
         if (($high && ($age > $high)) or ($age < $low)) {
             push @errors, 'ERROR_age_limitations';
             $template->param( age_low => $low);

